Author: Florian Dehau <work@fdehau.com>
Origin: commit:650e35ea213f17ce455a976dacfbdadd07059940
Description: Bump tui from 0.8.0 to 0.9.1
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -56,7 +56,7 @@ version = "0.8"
 version = "1.5.4"
 
 [dependencies.tui]
-version = "0.8.0"
+version = "0.9.0"
 
 [dependencies.unicode-width]
 version = "0.1.7"
--- a/src/ui/mod.rs
+++ b/src/ui/mod.rs
@@ -9,7 +9,7 @@ use rspotify::spotify::senum::RepeatStat
 use tui::backend::Backend;
 use tui::layout::{Constraint, Direction, Layout, Rect};
 use tui::style::{Color, Modifier, Style};
-use tui::widgets::{Block, Borders, Gauge, Paragraph, Row, SelectableList, Table, Text, Widget};
+use tui::widgets::{Block, Borders, Gauge, List, ListState, Paragraph, Row, Table, Text};
 use tui::Frame;
 use util::{
     create_artist_string, display_track_progress, get_color, get_percentage_width,
@@ -82,7 +82,7 @@ where
         .iter()
         .map(|item| Row::StyledData(item.iter(), gray));
 
-    Table::new(header.iter(), rows)
+    let help_menu = Table::new(header.iter(), rows)
         .block(
             Block::default()
                 .borders(Borders::ALL)
@@ -96,8 +96,8 @@ where
             Constraint::Length(50),
             Constraint::Length(40),
             Constraint::Length(20),
-        ])
-        .render(f, chunks[0]);
+        ]);
+    f.render_widget(help_menu, chunks[0]);
 }
 
 pub fn draw_input_and_help_box<B>(f: &mut Frame<B>, app: &App, layout_chunk: Rect)
@@ -117,16 +117,15 @@ where
     );
 
     let input_string: String = app.input.iter().collect();
-    Paragraph::new([Text::raw(&input_string)].iter())
-        .style(Style::default().fg(Color::Yellow))
-        .block(
-            Block::default()
-                .borders(Borders::ALL)
-                .title("Search")
-                .title_style(get_color(highlight_state))
-                .border_style(get_color(highlight_state)),
-        )
-        .render(f, chunks[0]);
+    let lines = [Text::raw(&input_string)];
+    let input = Paragraph::new(lines.iter()).block(
+      Block::default()
+       .borders(Borders::ALL)
+       .title("Search")
+       .title_style(get_color(highlight_state))
+       .border_style(get_color(highlight_state)),
+    );
+    f.render_widget(input, chunks[0]);
 
     let block = Block::default()
         .title("Help")
@@ -134,10 +133,11 @@ where
         .border_style(Style::default().fg(Color::Gray))
         .title_style(Style::default().fg(Color::Gray));
 
-    Paragraph::new([Text::raw("Type ?")].iter())
+    let lines = [Text::raw("Type ?")];
+    let help = Paragraph::new(lines.iter())
         .block(block)
-        .style(Style::default().fg(Color::Gray))
-        .render(f, chunks[1]);
+        .style(Style::default().fg(Color::Gray));
+    f.render_widget(help, chunks[1]);
 }
 
 pub fn draw_main_layout<B>(f: &mut Frame<B>, app: &App)
@@ -652,12 +652,12 @@ where
                 current_route.hovered_block == ActiveBlock::PlayBar,
             );
 
-            Block::default()
+            let title_block = Block::default()
                 .borders(Borders::ALL)
                 .title(&title)
                 .title_style(get_color(highlight_state))
-                .border_style(get_color(highlight_state))
-                .render(f, layout_chunk);
+                .border_style(get_color(highlight_state));
+            f.render_widget(title_block, layout_chunk);
 
             let track_name = if app
                 .liked_song_ids_set
@@ -668,25 +668,25 @@ where
                 track_item.name.clone()
             };
 
-            Paragraph::new(
-                [Text::styled(
-                    create_artist_string(&track_item.artists),
-                    Style::default().fg(Color::White),
-                )]
-                .iter(),
-            )
-            .style(Style::default().fg(Color::White))
-            .block(
+            let lines = [Text::styled(
+              create_artist_string(&track_item.artists),
+              Style::default().fg(Color::White),
+            )];
+            let artist = Paragraph::new(lines.iter())
+              .style(Style::default().fg(Color::White))
+              .block(
                 Block::default().title(&track_name).title_style(
-                    Style::default()
-                        .fg(Color::LightCyan)
-                        .modifier(Modifier::BOLD),
+                 Style::default()
+                  .fg(Color::LightCyan)
+                  .modifier(Modifier::BOLD),
                 ),
-            )
-            .render(f, chunks[0]);
+              );
+            f.render_widget(artist, chunks[0]);
             let perc = get_track_progress_percentage(app.song_progress_ms, track_item.duration_ms);
 
-            Gauge::default()
+            let song_progress_label =
+            display_track_progress(app.song_progress_ms, track_item.duration_ms);
+            let song_progress = Gauge::default()
                 .block(Block::default().title(""))
                 .style(
                     Style::default()
@@ -695,11 +695,8 @@ where
                         .modifier(Modifier::ITALIC | Modifier::BOLD),
                 )
                 .percent(perc)
-                .label(&display_track_progress(
-                    app.song_progress_ms,
-                    track_item.duration_ms,
-                ))
-                .render(f, chunks[1]);
+                .label(&song_progress_label);
+            f.render_widget(song_progress, chunks[1]);
         }
     }
 }
@@ -744,7 +741,7 @@ Hint: a playback device must be either a
         ))
     }
 
-    Paragraph::new(playing_text.iter())
+    let playing_paragraph = Paragraph::new(playing_text.iter())
         .wrap(true)
         .style(Style::default().fg(Color::White))
         .block(
@@ -753,8 +750,8 @@ Hint: a playback device must be either a
                 .title("Error")
                 .title_style(Style::default().fg(Color::Red))
                 .border_style(Style::default().fg(Color::Red)),
-        )
-        .render(f, chunks[0]);
+            );
+    f.render_widget(playing_paragraph, chunks[0]);
 }
 
 fn draw_home<B>(f: &mut Frame<B>, app: &App, layout_chunk: Rect)
@@ -773,12 +770,12 @@ where
         current_route.hovered_block == ActiveBlock::Home,
     );
 
-    Block::default()
+    let welcome = Block::default()
         .title("Welcome!")
         .borders(Borders::ALL)
         .title_style(get_color(highlight_state))
-        .border_style(get_color(highlight_state))
-        .render(f, layout_chunk);
+        .border_style(get_color(highlight_state));
+    f.render_widget(welcome, layout_chunk);
 
     let changelog = include_str!("../../CHANGELOG.md").to_string();
 
@@ -798,18 +795,18 @@ where
     ];
 
     // Contains the banner
-    Paragraph::new(top_text.iter())
+    let top_text = Paragraph::new(top_text.iter())
         .style(Style::default().fg(Color::White))
-        .block(Block::default())
-        .render(f, chunks[0]);
+        .block(Block::default());
+    f.render_widget(top_text, chunks[0]);
 
     // CHANGELOG
-    Paragraph::new(bottom_text.iter())
+    let bottom_text = Paragraph::new(bottom_text.iter())
         .style(Style::default().fg(Color::White))
         .block(Block::default())
         .wrap(true)
-        .scroll(app.home_scroll)
-        .render(f, chunks[1]);
+        .scroll(app.home_scroll);
+    f.render_widget(bottom_text, chunks[1]);
 }
 
 fn draw_not_implemented_yet<B>(
@@ -834,11 +831,11 @@ fn draw_not_implemented_yet<B>(
 
     let text = vec![Text::raw("Not implemented yet!")];
 
-    Paragraph::new(text.iter())
+    let not_implemented = Paragraph::new(text.iter())
         .style(Style::default().fg(Color::White))
         .block(display_block)
-        .wrap(true)
-        .render(f, layout_chunk);
+        .wrap(true);
+    f.render_widget(not_implemented, layout_chunk);
 }
 
 fn draw_artist_albums<B>(f: &mut Frame<B>, app: &App, layout_chunk: Rect)
@@ -880,14 +877,14 @@ where
         .margin(5)
         .split(f.size());
 
-    let device_instructions = vec![
-        "To play tracks, please select a device.",
-        "Use `j/k` or up/down arrow keys to move up and down and <Enter> to select",
-        "Your choice here will be cached so you can jump straight back in when you next open `spotify-tui`.",
-        "You can change the playback device at any time by pressing `d`.",
+    let device_instructions = [
+         Text::raw("To play tracks, please select a device."),
+         Text::raw("Use `j/k` or up/down arrow keys to move up and down and <Enter> to select"),
+         Text::raw("Your choice here will be cached so you can jump straight back in when you next open `spotify-tui`."),
+         Text::raw("You can change the playback device at any time by pressing `d`."),
     ];
 
-    Paragraph::new([Text::raw(device_instructions.join("\n"))].iter())
+    let instructions = Paragraph::new(device_instructions.iter())
         .style(Style::default().fg(Color::White))
         .wrap(true)
         .block(
@@ -895,27 +892,25 @@ where
                 .borders(Borders::NONE)
                 .title("Welcome to spotify-tui!")
                 .title_style(Style::default().fg(Color::Cyan).modifier(Modifier::BOLD)),
-        )
-        .render(f, chunks[0]);
+            );
+    f.render_widget(instructions, chunks[0]);
 
-    let no_device_message = vec!["No devices found: Make sure a device is active".to_string()];
+    let no_device_message = vec![Text::raw("No devices found: Make sure a device is active")];
 
-    let items = match &app.devices {
+    let items: Box<dyn Iterator<Item = Text>> = match &app.devices {
         Some(items) => {
             if items.devices.is_empty() {
-                no_device_message
+                Box::new(no_device_message.into_iter())
             } else {
-                items
-                    .devices
-                    .iter()
-                    .map(|device| device.name.to_owned())
-                    .collect()
+                Box::new(items.devices.iter().map(|device| Text::raw(&device.name)))
             }
         }
-        None => no_device_message,
+        None => Box::new(no_device_message.into_iter()),
     };
 
-    SelectableList::default()
+    let mut state = ListState::default();
+    state.select(app.selected_device_index);
+    let list = List::new(items)
         .block(
             Block::default()
                 .title("Devices")
@@ -923,15 +918,13 @@ where
                 .title_style(Style::default().fg(Color::LightCyan))
                 .border_style(Style::default().fg(Color::Gray)),
         )
-        .items(&items)
         .style(Style::default().fg(Color::White))
-        .select(app.selected_device_index)
         .highlight_style(
             Style::default()
                 .fg(Color::LightCyan)
                 .modifier(Modifier::BOLD),
-        )
-        .render(f, chunks[1]);
+         );
+    f.render_stateful_widget(list, chunks[1], &mut state);
 }
 
 pub fn draw_album_list<B>(f: &mut Frame<B>, app: &App, layout_chunk: Rect)
@@ -1072,7 +1065,10 @@ fn draw_selectable_list<B, S>(
     B: Backend,
     S: std::convert::AsRef<str>,
 {
-    SelectableList::default()
+    let mut state = ListState::default();
+    state.select(selected_index);
+
+    let list = List::new(items.iter().map(|i| Text::raw(i.as_ref())))
         .block(
             Block::default()
                 .title(title)
@@ -1080,11 +1076,9 @@ fn draw_selectable_list<B, S>(
                 .title_style(get_color(highlight_state))
                 .border_style(get_color(highlight_state)),
         )
-        .items(items)
         .style(Style::default().fg(Color::White))
-        .select(selected_index)
-        .highlight_style(get_color(highlight_state).modifier(Modifier::BOLD))
-        .render(f, layout_chunk);
+        .highlight_style(get_color(highlight_state).modifier(Modifier::BOLD));
+   f.render_stateful_widget(list, layout_chunk, &mut state);
 }
 
 fn draw_table<B>(
@@ -1163,7 +1157,7 @@ fn draw_table<B>(
         .map(|h| Constraint::Length(h.width))
         .collect::<Vec<tui::layout::Constraint>>();
 
-    Table::new(header.items.iter().map(|h| h.text), rows)
+    let table = Table::new(header.items.iter().map(|h| h.text), rows)
         .block(
             Block::default()
                 .borders(Borders::ALL)
@@ -1173,6 +1167,6 @@ fn draw_table<B>(
                 .border_style(get_color(highlight_state)),
         )
         .style(Style::default().fg(Color::White))
-        .widths(&widths)
-        .render(f, layout_chunk);
+        .widths(&widths);
+    f.render_widget(table, layout_chunk);
 }
